assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java

```diff
@@ -13554,9 +13565,21 @@ private void uploadAll_updateSubmissions(SessionState state,
                                              Map<String, UploadGradeWrapper> submissionTable, Set<AssignmentSubmission> submissions, Assignment assignment) {
         if (assignment != null && submissions != null) {
             for (AssignmentSubmission submission : submissions) {
-                if (submissionTable.containsKey(submission.getSubmitters().toArray(new AssignmentSubmissionSubmitter[0])[0].getSubmitter())) {
+                String eid;
+                if (assignment.getIsGroup()) {
+                    eid = submission.getGroupId();
+                } else {
+                    Optional<AssignmentSubmissionSubmitter> submitter = submission.getSubmitters().stream().findAny();
+                    if (submitter.isPresent()) {
+                        eid = submitter.get().getSubmitter();
+                    } else {
+                        log.warn("Upload while updating submission missing submitter for submission {}", submission.getId());
+                        continue;
+                    }
+                }
+                if (submissionTable.containsKey(eid)) {
```

Comment:
okay to call containsKey with a null eid?

Reply:
Yeah it's better to just call get and check, I wasn't actually changing the logic there!

Reply:
no i mean eid is null ....

Reply:
no it should never be null

Reply:
You're saying the same thing, essentially. Earle is just pointing out that it's more efficient to call `collection.get(eid)` then check the result for `null`.

Yes, there's an NPE possibility here, and yes it could be written more efficiently as Earle pointed out.

Reply:
I think `eid` can be null in this situation, given the code flow.

Reply:
i cleaned it up a little more and made it more visible, I think the question was around the uninitialized eid? If so the original code was fine as it covered all cases except for the chance that the group id would be missing however if getIsGroup is true and there is no group id we have bigger problems!

Reply:
okay got it

Reply:
this code is a perfectly valid initializer:
```
String text;              // this line is doesn't perform any initialization and if it did it would be redundant
if (something) {
  text = "something true";
} else {
  text = "something false";
}
```

Reply:
agree your example is perfectly valid. but in the actual code the else has an extra if statement in it

Reply:
correct but the otherside of that if statement was a `continue;` which ends this iteration, also in  the current form the continue is no longer needed as it is the last line in the loop!